runtime/j9jitbuilder/JavaMethodCallbackGen.cpp

```diff
@@ -19,23 +19,272 @@
+		rc == OOM;
+	} else {
+		DefineLine(LINETOSTR(__LINE__));
+		DefineFile(__FILE__);
+
+		for (int i = 0; i < _paramTypeLen; i++) {
+			snprintf((char*)(_paramIdentifiers + i), sizeof(U_16),"%d\0", i);
+			DefineParameter((char*)(_paramIdentifiers + i), getILTypeFromJ9Class(_paramTypes[i]));
+		}
+
+		DefineName("JavaMethodCallbackGen");
+		DefineReturnType(getILTypeFromJ9Class(returnType));
+
+		DefineFunction((char *)getCallInFunctionName(),
+		                  (char *)"jni.h",
+		                  getCallInFunctionLineNumber(),
+		                  getCallInFunctionAddress(),
+						  getILTypeFromJ9Class(_returnType),
+		                  3,
+						  Address,
+						  Address,
+						  Address);
+
+		if (_paramTypeLen > 0) {
+			_dictionary->DefineUnion("jvalue");
+			_dictionary->UnionField("jvalue", "Z", _dictionary->toIlType<uint8_t>());
+			_dictionary->UnionField("jvalue", "B", Int8);
+			_dictionary->UnionField("jvalue", "C", _dictionary->toIlType<uint16_t>());
+			_dictionary->UnionField("jvalue", "S", Int16);
+			_dictionary->UnionField("jvalue", "I", Int32);
+			_dictionary->UnionField("jvalue", "J", Int64);
+			_dictionary->UnionField("jvalue", "F", Float);
+			_dictionary->UnionField("jvalue", "D", Double);
+			_dictionary->CloseUnion("jvalue");
+		}
+	}
+
+	return rc;
+}
+
+void
+JavaMethodCallbackGen::destroy()
 {
-	DefineLine(LINETOSTR(__LINE__));
-	DefineFile(__FILE__);
-	DefineName("JavaMethodCallbackGen");
+	PORT_ACCESS_FROM_VMC(_vmThread);
+	j9mem_free_memory(_paramIdentifiers);
 }
 
 bool
 JavaMethodCallbackGen::buildIL()
 {
+	if (_paramTypeLen > 0) {
+		TR::IlType *jValueType = _dictionary->LookupUnion("jvalue");
+		TR::IlType *pJvalue = _dictionary->PointerTo(jValueType);
+
+		/* create array of jvalues (length of parameters) on stack */
+		Store("parameters",
+				CreateLocalArray(_paramTypeLen, jValueType));
+
+		for (int i = 0; i < _paramTypeLen; i++) {
+			/* get pointer to jvalue at parameter index */
+			Store("param", IndexAt(pJvalue,
+			         Load("parameters"),
+			         ConstInt32(i)));
+
+			/* fill jvalue with zero first */
+			StoreIndirect("jvalue", "J",
+								Load("param"),
+								ConstInt64(0));
+
+			/* load parameter into jvalue */
+			StoreIndirect("jvalue", getSignatureFromJ9Class(_paramTypes[i]),
+					Load("param"),
+					Load((char*)(_paramIdentifiers + i)));
+
+		}
+
+
+	} else {
+		Store("parameters", ConstAddress(NULL));
+	}
+
+	if (getILTypeFromJ9Class(_returnType) == NoType) {
+		Call(getCallInFunctionName(), 4, ConstAddress((void*)_vmThread), ConstAddress((void*) _anonInstance), ConstAddress(_methodID), Load("parameters"));
+		Return();
+	} else {
+		Return(Call(getCallInFunctionName(), 4, ConstAddress((void*)_vmThread), ConstAddress((void*) _anonInstance), ConstAddress(_methodID), Load("parameters")));
+	}
+
 	return true;
 }
+
+void*
+JavaMethodCallbackGen::getCallInFunctionAddress()
+{
+	void *functionAddress = NULL;
+	if (_returnType == _vm->voidReflectClass) {
```

Comment:
Is there a relationship between the index from the first `_vm->XXXReflectClass` to the index in the `_vmThread->functions->CallXXXMethodA` so we can do a bit of math to index into the tables?

Reply:
It is possible to do so. But, we would have to add objectReflectClass to J9JavaVM and reorder the rest of the reflect fields for that to work.